### PR TITLE
Wrap ec2 methods into functions and retry when requests are throttled

### PR DESCRIPTION
**Pull request by CyberDem0n**
In busy accounts chances of throttling aws api requests are very high

### CODE REVIEW

File: runtime/opt/taupage/init.d/10-prepare-disks.py [hunk 1]
```diff
@@ -29,14 +30,52 @@ def zone():
     return boto.utils.get_instance_metadata()['placement']['availability-zone']
 
 
+def retry(func):
+    def wrapped(*args, **kwargs):
+        count = 0
+        while True:
+            try:
+                return func(*args, **kwargs)
+            except boto.exceptions.BotoServerError as e:
```
Comment (a1exsh) [inline runtime/opt/taupage/init.d/10-prepare-disks.py]: singular: `boto.exception`. http://boto.cloudhackers.com/en/latest/ref/boto.html#module-boto.exception

File: runtime/opt/taupage/init.d/10-prepare-disks.py [hunk 2]
```diff
@@ -29,14 +30,52 @@ def zone():
...
+                if str(e.error_code) not in ('Throttling', 'RequestLimitExceeded'):
+                    raise
+                logging.info('Throttling AWS API requests...')
+
+            if count >= 10:
+                break
+            time.sleep(2 ** count * 0.5)
```
Comment (a1exsh) [inline runtime/opt/taupage/init.d/10-prepare-disks.py]: So this sleeps up to roughly 1 second in total?  Not sure that's reliable enough.

Comment (a1exsh) [inline runtime/opt/taupage/init.d/10-prepare-disks.py]: Scratch that, it's not in milliseconds :-p  Should be OK.

File: runtime/opt/taupage/init.d/10-prepare-disks.py (hunk 1 above)
Comment (CyberDem0n) [inline runtime/opt/taupage/init.d/10-prepare-disks.py]: Fixed, thank you

File: runtime/opt/taupage/init.d/10-prepare-disks.py [hunk 3]
```diff
@@ -213,20 +243,20 @@ def should_format_volume(region, partition, erase_on_boot):
     #
     if erase_on_boot is None:
         ec2 = ec2_client(region)
-        volumes = list(ec2.get_all_volumes(filters={
+        volumes = list(get_all_volumes(ec2, filters={
```
Comment (a1exsh) [inline runtime/opt/taupage/init.d/10-prepare-disks.py]: It should still work like this, but for consistency maybe remove `filters=` part?

File: runtime/opt/taupage/init.d/10-prepare-disks.py [hunk 4]
```diff
@@ -213,20 +243,20 @@ def should_format_volume(region, partition, erase_on_boot):
...
             'attachment.instance-id': instance_id(),
             'attachment.device': partition}))
         if volumes:
             try:
                 volume_id = volumes[0].id
                 logging.info("%s: volume_id=%s", partition, volume_id)
 
-                tags = ec2.get_all_tags(filters={
+                tags = get_all_tags(ec2, filters={
```
Comment (a1exsh) [inline runtime/opt/taupage/init.d/10-prepare-disks.py]: Again `filters=`.

**Comment by a1exsh** [comment]
There are two other places in Taupage where boto APIs are used.  Does it make sense to improve their reliability as well?

https://github.com/zalando-stups/taupage/search?l=Python&q=boto&type=&utf8=%E2%9C%93

**Comment by CyberDem0n** [comment]
Definitely

**Comment by a1exsh** [comment]
:+1:

**Comment by tuxlife** [comment]
:+1:

**Pull request merged**